### PR TITLE
Fix Task Progress

### PR DESCRIPTION
**Pull request by theroyalblah**
What I thought would take 30 minutes instead took 3 days.

The solution to calculating the progress is disastrously heavy handed and confusing. Honestly, good luck figuring out the mess that I had to create to make this work.

This is very heavy on fetches and might be able to be reduced, but I don't think without some major reworks this could be improved much.

Anyhow, it doesn't work as well as I'd like. For one, dependent on the speed of the database, the progress on the left may or may not update. Run the db locally and it works like a charm, otherwise your performance will vary. Please test this thoroughly and look for bugs, I expect some to exist.

image

### CODE REVIEW

**Comment by NickCampagna** [comment]
> I don't have anything to say about the code, and the actual progress increasing looks great! However I did seem to run into problems that I'm not sure are caused by this code or by something else, but I'll bring it up because I'm just not sure.
> When I create a project, and then create a task for that project, and then create a subtask, (I'm not sure if this is particularly because I change to complete or if I add more subtasks), the project will disappear from the project sidebar and the overall head/parent will shift to the task instead. I've been trying to figure out the Exact situation of which it occurs at, but I'm not completely sure(the db crashed while I was testing, so I don't know the Exact cause yet, but I was able to replicate the problem a number of times).
> I just wanted to bring up this problem b/c its a Big one, even if I'm unsure if this pr caused it or if it was a problem we just never noticed before.

Whoop, forgot to replace the standard "Delete" text in the other fonts/themes with {this.deleteButtonText} . Fixed that, and changed the "CONFIRM DELETION" text to all-caps "DELETE" to allow the user to confirm while preventing the button from expanding further.

**Comment by JustinLaGree** [comment]
@ignatyevE I would recommend we write this up as an issue and allow the pull request through unless we know that this PR is causing the bug.

File: src/components/taskView.tsx [hunk 1]
```diff
@@ -577,32 +578,36 @@ export class TaskView extends React.Component<TaskViewProps, TaskViewState>{
...
+            }
         }
         // TODO
         // should be user from google oauth
-        const updatedTask = { owner: this.owner, parentId: this.props.parentId, projectId: this.props.projectId, title: this.state.name, status: this.state.taskStatus, assignedTo: this.state.assignedState, progress: this.state.completion, deadline: this.state.dueDate, description: this.state.description };
+        const updatedTask = { owner: this.owner, parentId: this.props.parentId, projectId: this.props.projectId, title: this.state.name, status: this.state.taskStatus, assignedTo: this.state.assignee, progress: completion, deadline: this.state.dueDate, description: this.state.description };
 
         const request = new UserHeaderHttpRequest(`/api/tasks/${this.props.taskID}`, this.props.userInfo, { "Content-Type" : "application/json" });
-        RetryableFetch.fetch_retry(request,
+        fetch(request,
```
Comment (JustinLaGree) [inline src/components/taskView.tsx:603]: Make this a retryable fetch.

File: src/components/projectPage.tsx [hunk 2]
```diff
@@ -204,6 +212,130 @@ export class ProjectPage extends React.Component<ProjectPageProps, { error: any,
 		this.changeHead(newHead, true);
 	}
 
+	// get the progress of all the children and return it as an array
+	private retrieveChildProgress = async (id: number): Promise<number[]> => {
+
+		let childProgress: number[] = [];
+		//get subtask progress
+		const request = new UserHeaderHttpRequest(`/api/subtasks/${id}`, this.props.userInfo);
+		await fetch(request)
```
Comment (JustinLaGree) [inline src/components/projectPage.tsx:221]: Make this a retryable fetch.

File: src/components/projectPage.tsx [hunk 3]
```diff
@@ -204,6 +212,130 @@ export class ProjectPage extends React.Component<ProjectPageProps, { error: any,
...
+
+		return childProgress;
+	}
+
+	// refresh fetch, called after a save. The head does not change, but names might have,
+	// so the data needs to be recollected.
+	private refreshFetch = () => {
+		let history = this.state.history;
+		const request = new UserHeaderHttpRequest(`/api/tasks/${this.state.head}`, this.props.userInfo);
+		fetch(request)
```
Comment (JustinLaGree) [inline src/components/projectPage.tsx:242]: Make this a retryable fetch.

File: src/components/projectPage.tsx [hunk 4]
```diff
@@ -204,6 +212,130 @@ export class ProjectPage extends React.Component<ProjectPageProps, { error: any,
...
+				}
+			);
+	}
+
+	private updateProgress = async (id: number, progress: number) => {
+
+		const updatedTask = { owner: this.state.task.owner, title: this.state.task.title, status: this.state.task.status, assignedTo: this.state.task.assignedTo, progress: progress, deadline: this.state.task.deadline, description: this.state.task.description };
+		//get subtask progress
+		const request = new UserHeaderHttpRequest(`/api/tasks/${id}`, this.props.userInfo, { 'Content-Type': 'application/json' });
+		await fetch(request,
```
Comment (JustinLaGree) [inline src/components/projectPage.tsx:327]: Make this a retryable fetch.

**Pull request merged**